experimental/devices/sn3218/sn3218.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2018 The Periph Authors. All rights reserved.
+// Use of this source code is governed under the Apache License, Version 2.0
+// that can be found in the LICENSE file.
+
+package sn3218
+
+import (
+	"errors"
+
+	"periph.io/x/periph/conn/i2c"
+)
+
+const (
+	i2cAddress             = 0x54
+	cmdEnableOutput        = 0x00
+	cmdSetBrightnessValues = 0x01
+	cmdEnableLeds          = 0x13
+	cmdUpdate              = 0x16
+	cmdReset               = 0x17
+)
+
+// Dev is a handler to sn3218 controller.
+type Dev struct {
+	i2c        i2c.Dev
+	states     [18]bool
+	brightness [18]byte
+}
+
+// New returns a handle to a SN3218 LED driver.
+func New(bus i2c.Bus) (Dev, error) {
+	d := i2c.Dev{Bus: bus, Addr: i2cAddress}
+	dev := Dev{}
+	dev.i2c = d
+	dev.reset()
+	return dev, nil
+}
+
+// Halt resets the registers and switches the driver off.
+func (d *Dev) Halt() error {
+	if err := d.Disable(); err != nil {
+		return err
+	}
+	return d.reset()
+}
+
+// Enable enables the SN3218.
+func (d *Dev) Enable() error {
+	_, err := d.i2c.Write([]byte{cmdEnableOutput, 0x01})
+	return err
+}
+
+// Disable disables the SN3218.
+func (d *Dev) Disable() error {
+	_, err := d.i2c.Write([]byte{cmdEnableOutput, 0x00})
+	return err
+}
+
+// GetState returns the state (on/off) and the brightness (0..255) of the
+// Channel 0..17.
+func (d *Dev) GetState(channel int) (bool, byte, error) {
+	if channel < 0 || channel >= 18 {
+		return false, 0, errors.New("Channel number out of range 0..17")
+	}
+	return d.states[channel], d.brightness[channel], nil
+}
+
+// Switch switched the channel (0..18) to state (on/off).
+func (d *Dev) Switch(channel int, state bool) error {
+	if channel < 0 || channel >= 18 {
+		return errors.New("Channel number out of range 0..17")
+	}
+	d.states[channel] = state
+	return d.updateStates()
+}
+
+// SetGlobalBrightness sets the brightness of all channels to the value (0..255).
+func (d *Dev) SetGlobalBrightness(value byte) {
+	for i := 0; i < 18; i++ {
+		d.brightness[i] = value
+	}
+	d.updateBrightness()
+}
+
+// SetBrightness sets the brightness of led (0..17) to value (0..255).
+func (d *Dev) SetBrightness(channel int, value byte) error {
+	if channel < 0 || channel >= 18 {
+		return errors.New("Chhannel number out of range 0..17")
+	}
+	d.brightness[channel] = value
+	d.updateBrightness()
+	return nil
+}
+
+// SwitchAll switches all channels accoring to the state (on/off).
+func (d *Dev) SwitchAll(state bool) {
+	for i := 0; i < 18; i++ {
+		d.states[i] = state
+	}
+	d.updateStates()
+}
+
+// Reset resets the registers to the default values.
+func (d *Dev) reset() error {
+	_, err := d.i2c.Write([]byte{cmdReset, 0xFF})
+	d.states = [18]bool{}
```

Comment:
I wonder if the compiler is able to optimize that, vs a loop.

Reply:
100 mio x [18]bool{} - Took: 41.262059ms
100 mio x loop - Took: 821.613264ms

```
package main

import (
        "fmt"
        "time"
)

func performance() {
        var on [18]bool
        start := time.Now()
        for i := 0; i < 100000000; i++ {
                on = [18]bool{}
        }
        elapsed := time.Since(start)
        fmt.Printf("100 mio x [18]bool{} - Took: %s\n", elapsed)

        start2 := time.Now()
        for i := 0; i < 100000000; i++ {
                for i := 0; i < 18; i++ {
                        on[i] = false
                }
        }
        elapsed2 := time.Since(start2)
        fmt.Printf("100 mio x loop - Took: %s\n", elapsed2)
        _ = on
}

func main() {
        for i := 0; i < 10; i++ {
                performance()
                fmt.Println("***************")
        }
}
```